### ISSUE TITLE
run on windows

### ISSUE THREAD

**Issue by ss880426** [sjtuplayer/anomalydiffusion#46]
Is there a way to run this model on Windows?
Thank you

**Comment by dirkjankroon** [comment]
Yes, currently I'm training the model in Windows, using CUDA 12.6 on a RTX4090 with Python 3.10

To make the code run I made the following tweaks:

In main.py, and train_mask.py I replaced:
trainer_config["accelerator"] = "ddp"" 
by   
trainer_config["gpus"] = 1
Because distributed-data-parallel with NCCL is not supported in Windows.

In main.py, and train_mask.py I replaced
signal.signal(signal.SIGUSR1, melk)
signal.signal(signal.SIGUSR2, divein)
by
signal.signal(signal.SIGINT, melk)
signal.signal(signal.SIGINT, divein)
Because SIGUSR1 is not available on Windows.

Also in imgaug.py in the folder ...Lib\site-packages\imgaug
I changed line 45-47 from
NP_FLOAT_TYPES = set(np.sctypes["float"])
NP_INT_TYPES = set(np.sctypes["int"])
NP_UINT_TYPES = set(np.sctypes["uint"])
To:
NP_FLOAT_TYPES = {np.float32, np.float64, np.float16}
NP_INT_TYPES = {np.int8, np.int16, np.int32, np.int64}
NP_UINT_TYPES = {np.uint8, np.uint16, np.uint32, np.uint64}
This is need to support Numpy 2.* or higher

Also in personalized.py, personalized_style.py and Isun.py
I replaced every instance of 
PIL.Image.LINEAR 
by
PIL.Image.BILINEAR
Because LINEAR does not exist any more in PILLOW

In run-mvtec.py 
Replaced
CUDA_VISIBLE_DEVICES=%d python
by
set CUDA_VISIBLE_DEVICES=%d & python
To set the environment variable correctly in Windows, and & to indicate an additional command

In embedding_manager.py (line 171) embedding_manager2.py (line 198), generate_mask.py (16), main.py (29)
        ... = torch.load(ckpt_path, map_location='cpu')
by
        ... = torch.load(ckpt_path, map_location='cpu', weights_only = False)


The files in the folder "taming" where not complete
I downloaded the files of this folder from "https://github.com/CompVis/taming-transformers/tree/master/taming"


As environment variables, I use:
CUDA_HOME=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.6
CUDA_PATH=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.6
path=C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.43.34808\bin\Hostx64\x64;C:\Python310;C:\Python310\Scripts;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.6\bin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.6\libnvvp;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.6\extras\CUPTI\lib64;C:\Windows\System32;C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\bin\Hostx64\x64;C:\Program Files\NVIDIA GPU Computing Toolkit\cudnn\bin;C:\Program Files\Microsoft Visual Studio\2022\Professional\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin;C:\Program Files\Microsoft Visual Studio\2022\Professional\Common7\IDE\CommonExtensions\Microsoft\CMake\Ninja;C:\Program Files\Microsoft Visual Studio\2022\Professional\Common7\IDE\VC\Linux\bin\ConnectionManagerExe;C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\vcpkg

Important is this one: C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\bin\Hostx64\x6
There is the CL.exe the microsoft visual studio compiler. If you don't have Microsoft Visual Studio 2022 installed, you can download the free Visual Studio 2022 community edition, or install the Windows SDK 

Visual Studio 2022 works only for CUDA 12.4 or higher, thus I installed 12.6
https://developer.nvidia.com/cuda-12-6-0-download-archive

Pytorch was installed with:
pip3 install torch torchvision --index-url https://download.pytorch.org/whl/cu126

Clip was installed with:
pip install git+https://github.com/openai/CLIP.git

I did not use conda as packet manager, but with pip Installed the following python packages:

Package                   Version
------------------------- ---------------
absl-py                   2.3.1
addict                    2.4.0
aiofiles                  24.1.0
aiohappyeyeballs          2.6.1
aiohttp                   3.12.14
aiosignal                 1.4.0
albumentations            1.3.0
altair                    5.5.0
annotated-types           0.7.0
antlr4-python3-runtime    4.8
anyio                     4.9.0
async-timeout             5.0.1
attrs                     25.3.0
basicsr                   1.4.2
blinker                   1.9.0
braceexpand               0.1.7
cachetools                6.1.0
certifi                   2025.7.14
charset-normalizer        3.4.2
click                     8.2.1
clip                      1.0
colorama                  0.4.6
contourpy                 1.3.2
cycler                    0.12.1
einops                    0.3.0
exceptiongroup            1.3.0
fastapi                   0.116.1
ffmpy                     0.6.0
filelock                  3.18.0
fonttools                 4.58.5
frozenlist                1.7.0
fsspec                    2025.5.1
ftfy                      6.3.1
future                    1.0.0
gitdb                     4.0.12
GitPython                 3.1.44
gradio                    3.16.2
grpcio                    1.73.1
h11                       0.16.0
httpcore                  1.0.9
httpx                     0.28.1
huggingface-hub           0.33.4
idna                      3.10
imageio                   2.37.0
imgaug                    0.4.0
importlib_metadata        8.7.0
Jinja2                    3.1.6
joblib                    1.5.1
jsonschema                4.24.0
jsonschema-specifications 2025.4.1
kiwisolver                1.4.8
kornia                    0.6.0
lazy_loader               0.4
lightning-utilities       0.14.3
linkify-it-py             2.0.3
lmdb                      1.7.2
Markdown                  3.8.2
markdown-it-py            3.0.0
MarkupSafe                3.0.2
matplotlib                3.10.3
mdit-py-plugins           0.4.2
mdurl                     0.1.2
mpmath                    1.3.0
multidict                 6.6.3
narwhals                  1.47.0
networkx                  3.4.2
ninja                     1.11.1.4
numpy                     2.2.6
omegaconf                 2.1.1
open-clip-torch           2.0.2
opencv-python             4.12.0.88
opencv-python-headless    4.12.0.88
orjson                    3.10.18
packaging                 25.0
pandas                    2.3.1
pillow                    11.3.0
pip                       25.1.1
prettytable               3.6.0
propcache                 0.3.2
protobuf                  3.20.3
pyarrow                   20.0.0
pycryptodome              3.23.0
pydantic                  2.11.7
pydantic_core             2.33.2
pydeck                    0.9.1
pyDeprecate               0.3.1
pydub                     0.25.1
Pygments                  2.19.2
Pympler                   1.1
pyparsing                 3.2.3
python-dateutil           2.9.0.post0
python-multipart          0.0.20
pytorch-lightning         1.5.1
pytz                      2025.2
pywin32                   311
PyYAML                    6.0.2
qudida                    0.0.4
referencing               0.36.2
regex                     2024.11.6
requests                  2.32.4
rich                      14.0.0
rpds-py                   0.26.0
safetensors               0.2.7
scikit-image              0.25.2
scikit-learn              1.7.0
scipy                     1.15.3
semver                    3.0.4
setuptools                57.4.0
shapely                   2.1.1
six                       1.17.0
smmap                     5.0.2
sniffio                   1.3.1
starlette                 0.47.1
streamlit                 1.12.1
sympy                     1.13.3
tb-nightly                2.20.0a20250715
tensorboard               2.20.0
tensorboard-data-server   0.7.2
test_tube                 0.7.5
threadpoolctl             3.6.0
tifffile                  2025.5.10
timm                      0.6.12
tokenizers                0.12.1
toml                      0.10.2
torch                     2.7.1+cu126
torchmetrics              1.7.4
torchvision               0.22.1+cu126
tornado                   6.5.1
tqdm                      4.67.1
transformers              4.19.2
typing_extensions         4.14.1
typing-inspection         0.4.1
tzdata                    2025.2
tzlocal                   5.3.1
uc-micro-py               1.0.3
urllib3                   2.5.0
uvicorn                   0.35.0
validators                0.35.0
watchdog                  6.0.0
wcwidth                   0.2.13
webdataset                0.2.5
websockets                15.0.1
Werkzeug                  3.1.3
yapf                      0.32.0
yarl                      1.20.1
zipp                      3.23.0
